cli/azd/cmd/container.go

```diff
@@ -161,26 +167,23 @@ func registerCommonDependencies(container *ioc.NestedContainer) {
 	})
 
 	container.RegisterSingleton(
-		func(azdContext *azdcontext.AzdContext, envFlags flagsWithEnv) (*environment.Environment, error) {
+		func(ctx context.Context,
+			azdContext *azdcontext.AzdContext,
+			envFlags flagsWithEnv,
+			console input.Console,
+			accountManager account.Manager,
+			userProfileService *azcli.UserProfileService) (*environment.Environment, error) {
 			if azdContext == nil {
 				return nil, azdcontext.ErrNoProject
 			}
 
 			environmentName := envFlags.EnvironmentName()
 			var err error
 
-			if environmentName == "" {
-				defaultEnvName, err := azdContext.GetDefaultEnvironmentName()
-				if err != nil {
-					return nil, err
-				}
-
-				environmentName = defaultEnvName
-			}
-
-			env, err := environment.GetEnvironment(azdContext, environmentName)
+			env, err := loadOrInitEnvironment(
+				ctx, &environmentName, azdContext, console, accountManager, userProfileService)
```

Comment:
Do you have thoughts on how we could address this? I have a similar concern with the DI flow, but if we want to support:

1. Automatically prompting for environment
1. Automatically injecting the correct tenant token

Then it would need part of the flow.

For `hook`, I would prefer the access pattern through `Lazy`, and retrying hooks on each `err`.

Reply:
Now that hooks will load a lazy environment I think it might be OK?  I will test out your branch and see what happens.

Reply:
@vhvb1989 we don't support hooks for `init` command right now.  The most complex case is the `up` command where the `env` doesn't exist at first but then after `init` is called environment is available.  Hooks are needed for up and the other commands like `infra create` and `deploy` that are called internally.

Reply:
There's a straightforward to switch the `Lazy` impl that `hooks` depends on. Currently, we would end up prompting the user before executing the hooks if environment isn't initialized, which isn't a bad thing.